### PR TITLE
xapid 377

### PR DESCRIPTION
**Pull request by pjain404**
I have created the MR against Master. It is basically all new code and huge MR. I have added design details in the README but I can also do a session explaining the design so that it helps while doing the code review. 

This plugin will be loaded as part of apid and cannot function independently as it needs the data apigeeSync populates in the DB. So I will create a separate MR for apid plugin to load the analytics plugin as well on startup once this code is merged to master.

Also, functionality is complete and tested manually but I am still working on adding unit test cases to have maximum code coverage. 

[Referernce: High Level Design Doc](https://docs.google.com/a/apigee.com/document/d/1dmgjr1vSCxrRVbKsPBsOTgzuri3vEVmz6OncOFazYYw/edit?usp=sharing)
[Referernce: API Spec](http://playground.apistudio.io/64e409ad-aebb-4bbc-977e-f0e0f22209d4/docs/)

@sramamoorthy73 @alexkhimich

### CODE REVIEW

**Comment by sramamoorthy73** [comment]
Impressive. Give me some time (today evening/night). Let me go through the code.

File: buffering_manager.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		completeFilePath := filepath.Join(newPath, fileName)
+		fw, err := createGzipFile(completeFilePath)
+		if err != nil {
+			return bucket{}, err
+		}
+
+		newBucket := bucket{DirName: dirName, FileWriter: fw}
+		bucketMap[ts] = newBucket
+
+		//Send event to close directory after endTime + 5 seconds to make sure all buffers are flushed to file
```
Comment (sramamoorthy73) [inline buffering_manager.go]: May be in the future, make this fudge time configurable.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# ApidAnalytics
```
Comment (pjain404) [inline README.md]: I ran go fmt, go vet and the race checker tool.
1. go fmt - fixed the formatting
2. The race checker tool found a data race related to the way I was creating cache. I fixed that already.
3. There is another one it finds in buffer manager while trying to write and close a file. It thinks the file handler will cause a data race. I ran through the logic again and I am sure that will not be an issue since I am managing access to that variable by waiting on a channel.

**Comment by pjain404** [comment]
Thanks for approving it. 
1. Regarding some additional negative test cases on the caching/channel - I am working on a separate branch to add more test cases for the whole plugin and will create a MR for it in this week.
2. I am sure of the merge process, will you merge it to master or I have to do it ? @sramamoorthy73

File: buffering_manager.go (hunk 1 above)
Comment (pjain404) [inline buffering_manager.go]: ACK
